dev/src/bulk-writer.ts

```diff
@@ -0,0 +1,655 @@
+  }
+
+  /**
+   * Return the first eligible batch that can hold a write to the provided
+   * reference, or creates one if no eligible batches are found.
+   */
+  private getEligibleBatch(ref: DocumentReference): BulkCommitBatch {
+    let batch = null;
+    let logWarning = this.batchQueue.length > 0;
+    for (const bulkBatch of this.batchQueue) {
+      if (!bulkBatch.refsInBatch.has(ref)) {
+        logWarning = false;
+      }
+      if (bulkBatch.canAddDoc(ref)) {
+        batch = bulkBatch;
+      }
+    }
+
+    if (batch === null) {
+      if (logWarning) {
+        console.warn(
+          '[BulkWriter]',
+          'Writing to the same document multiple times will slow down BulkWriter. ' +
+            'Write to unique documents in order to maximize throughput.'
+        );
+      }
+      return this.createNewBatch();
+    } else {
+      return batch;
+    }
+  }
+
+  private createNewBatch(): BulkCommitBatch {
+    const newBatch = new BulkCommitBatch(
+      this.getId(),
+      this.firestore.batch(),
+      this.maxBatchSize
+    );
+
+    this.batchQueue.push(newBatch);
+    return newBatch;
+  }
+
+  private getId(): number {
+    const id = this.currentId;
+    if (this.currentId === Number.MAX_SAFE_INTEGER) {
+      this.currentId = 0;
+    } else {
+      this.currentId++;
+    }
+    return id;
+  }
+
+  /**
+   * Attempts to send batches starting from the front of the BatchQueue until a
+   * batch cannot be sent.
+   *
+   * After a batch is complete, try sending batches again.
+   */
+  private sendBatches(): void {
+    const unsentBatches = this.batchQueue.filter(
+      batch => batch.sendTime === null
+    );
+
+    let index = 0;
+    while (
+      unsentBatches.length > index &&
+      this.isBatchSendable(unsentBatches[index])
+    ) {
+      const batch = unsentBatches[index];
+      batch.refsInBatch.forEach(batch => {
+        this.refsInFlight.add(batch);
+      });
+
+      batch.bulkCommit().then(results => {
+        batch.resolveResults(results);
+
+        const batchIndex = this.batchQueue.indexOf(batch);
+        assert(batchIndex !== -1, 'The batch should be in the BatchQueue');
+        this.batchQueue.splice(batchIndex, 1);
+
+        // Remove references that are no longer in flight.
+        batch.refsInBatch.forEach(batch => {
+          this.refsInFlight.delete(batch);
+        });
+
+        this.sendBatches();
+      });
+
+      index++;
+    }
+  }
+
+  /**
+   * Checks that the provided batch is sendable. To be sendable, a batch must:
+   * (1) be marked as pending
+   * (2) not write to references that are currently in flight
+   * (3) stays under the operations rate limit if sent
+   */
+  private isBatchSendable(batch: BulkCommitBatch): boolean {
+    if (!batch.pending) {
+      return false;
+    }
+
+    for (const ref of batch.refsInBatch) {
+      if (this.refsInFlight.has(ref)) {
+        return false;
+      }
+    }
+
+    if (this.options && !this.options.throttlingEnabled) {
```

Comment:
You should copy `throttlingEnabled` into local state, otherwise the user can modify  `throttlingEnabled` after the fact.